use-cases/meeting-scheduling/src/main/java/org/acme/meetingschedule/solver/MeetingSchedulingConstraintProvider.java

```diff
@@ -0,0 +1,224 @@
+package org.acme.meetingschedule.solver;
+
+import static ai.timefold.solver.core.api.score.stream.Joiners.equal;
+import static ai.timefold.solver.core.api.score.stream.Joiners.filtering;
+import static ai.timefold.solver.core.api.score.stream.Joiners.greaterThan;
+import static ai.timefold.solver.core.api.score.stream.Joiners.lessThan;
+import static ai.timefold.solver.core.api.score.stream.Joiners.overlapping;
+
+import ai.timefold.solver.core.api.score.stream.Constraint;
+import ai.timefold.solver.core.api.score.stream.ConstraintFactory;
+import ai.timefold.solver.core.api.score.stream.ConstraintProvider;
+
+import org.acme.meetingschedule.domain.Attendance;
+import org.acme.meetingschedule.domain.MeetingAssignment;
+import org.acme.meetingschedule.domain.PreferredAttendance;
+import org.acme.meetingschedule.domain.RequiredAttendance;
+import org.acme.meetingschedule.domain.Room;
+import org.acme.meetingschedule.domain.TimeGrain;
+
+public class MeetingSchedulingConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+                roomConflict(constraintFactory),
+                avoidOvertime(constraintFactory),
+                requiredAttendanceConflict(constraintFactory),
+                requiredRoomCapacity(constraintFactory),
+                startAndEndOnSameDay(constraintFactory),
+                requiredAndPreferredAttendanceConflict(constraintFactory),
+                preferredAttendanceConflict(constraintFactory),
+                doMeetingsAsSoonAsPossible(constraintFactory),
+                oneBreakBetweenConsecutiveMeetings(constraintFactory),
+                overlappingMeetings(constraintFactory),
+                assignLargerRoomsFirst(constraintFactory),
+                roomStability(constraintFactory)
+        };
+    }
+
+    // ************************************************************************
+    // Hard constraints
+    // ************************************************************************
+
+    public Constraint roomConflict(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEachUniquePair(MeetingAssignment.class,
+                equal(MeetingAssignment::getRoom),
+                overlapping(MeetingAssignment::getGrainIndex,
+                        assignment -> assignment.getGrainIndex() +
+                                assignment.getMeeting().getDurationInGrains()))
```

Comment:
Don't we have a method for this? If not, let's add it. It's used very often. Please check the rest of the constraint provider too.

Reply:
We have a slightly different method for calculating the index. Now, we have a method for calculating the duration.